File: fp-in-clojure/src/fp_in_clojure/exercises/getting_started/anonymous_functions.clj
;; 関数型プログラミング(FP)では関数を取り回すことがよくあるため、
;; 名前を付けること *なく* 関数を組み立てるシンタックスがあると便利だ
(ns fp-in-clojure.exercises.getting-started.anonymous-functions
  (:require
   [fp-in-clojure.exercises.getting-started.my-program :refer [abs factorial format-result]]))

;; 無名関数の例:
(defn -main [& _]
  (println (format-result "absolute value" -42 abs))
  (println (format-result "factorial" 7 factorial))
  (println (format-result "increment" 7 (fn [x] (inc x))))
  ;; NOTE: `#(f %)` や `#(g %1 %2 %3)` のように使える無名関数(ラムダ式)の略記法
  ;; ref. https://clojure.org/guides/weird_characters#_anonymous_function
  (println (format-result "increment2" 7 #(inc %)))
  (println (format-result "increment3" 7 inc))
  (println (format-result "increment4" 7 (fn [x] (let [r (inc x)] r)))))

(comment
  (with-out-str
    (-main))
  )

File: fp-in-haskell/src/FpInHaskell/Exercises/GettingStarted/AnonymousFunctions.hs
module FpInHaskell.Exercises.GettingStarted.AnonymousFunctions (
    printAnonymousFunctions,
) where

import FpInHaskell.Exercises.GettingStarted.MyProgram (abs, factorial, formatResult)
import Prelude hiding (abs)

-- 関数型プログラミングでは関数を取り回すことが多いため、
-- 名前を付けることなく関数を組み立てる構文があると便利だ
-- 無名関数の例:
printAnonymousFunctions :: IO ()
printAnonymousFunctions = do
    putStrLn (formatResult "absolute value" (-42) abs)
    putStrLn (formatResult "factorial" 7 factorial)
    putStrLn (formatResult "increment" 7 (\x -> x + 1))
    putStrLn (formatResult "increment2" 7 (\x -> x + 1))
    putStrLn (formatResult "increment3" 7 (\x -> x + 1))
    putStrLn (formatResult "increment4" 7 (+ 1)) -- `(+ 1)` は `+` の部分適用。`\x -> x + 1` と同じ
    putStrLn (formatResult "increment5" 7 (\x -> let r = x + 1 in r))
